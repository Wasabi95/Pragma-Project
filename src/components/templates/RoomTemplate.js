//src/components/templates/RoomTemplate.js
import React from 'react';
import Button from '../atoms/Button';
// We don't need to import the Badge atom here as we are using the bootstrap class directly.

function RoomTemplate({ roomHeaderInfo, participantsPanel, moderatorControls, storyPanel, votingPanel }) {
  // 1. Destructure the new userRole prop
  const { roomName, userName, isModerator, userRole, onCopyLink } = roomHeaderInfo;
  
  return (
    <div>
      <div className="d-flex justify-content-between align-items-center mb-3">
        <h2>Room: {roomName}</h2>
        <div>
          <span>Welcome, <strong>{userName}</strong></span>
          
          {/* 2. Add the rendering logic for the role badge */}
          {userRole && (
            <span className="badge bg-info text-dark ms-2">
              {userRole === 'PLAYER' ? 'Jugador' : 'Espectador'}
            </span>
          )}

          {/* The existing Moderator badge logic remains */}
          {isModerator && <span className="badge bg-primary ms-2">Moderator</span>}
        </div>
      </div>
      
      {/* ... (rest of the template remains the same) ... */}
      <Button onClick={onCopyLink} className="btn-outline-primary mb-4">
        ðŸ“‹ Copy Room Link
      </Button>

      <div className="row">
        <div className="col-md-4">
          {participantsPanel}
          {isModerator && moderatorControls}
        </div>
        <div className="col-md-8">
          {storyPanel}
          {votingPanel}
        </div>
      </div>
    </div>
  );
}

export default RoomTemplate;