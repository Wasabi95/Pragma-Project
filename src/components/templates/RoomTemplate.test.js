// src/components/templates/RoomTemplate.test.js
// src/components/templates/RoomTemplate.test.js
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import RoomTemplate from './RoomTemplate';

describe('RoomTemplate Component', () => {
  const mockParticipantsPanel = <div data-testid="participants">Participants</div>;
  const mockModeratorControls = <div data-testid="moderator">Moderator Controls</div>;
  const mockStoryPanel = <div data-testid="story">Story Panel</div>;
  const mockVotingPanel = <div data-testid="voting">Voting Panel</div>;

  const baseProps = {
    roomHeaderInfo: {
      roomName: 'Test Room',
      userName: 'Test User',
      isModerator: false,
      userRole: 'PLAYER',
      onCopyLink: jest.fn()
    },
    participantsPanel: mockParticipantsPanel,
    moderatorControls: mockModeratorControls,
    storyPanel: mockStoryPanel,
    votingPanel: mockVotingPanel
  };

  test('renders basic room information', () => {
    render(<RoomTemplate {...baseProps} />);
    
    expect(screen.getByText('Room: Test Room')).toBeInTheDocument();
    expect(screen.getByText('Welcome,')).toBeInTheDocument();
    expect(screen.getByText('Test User')).toBeInTheDocument();
  });

  test('shows player badge when user is PLAYER', () => {
    render(<RoomTemplate {...baseProps} />);
    expect(screen.getByText('Jugador')).toBeInTheDocument();
    expect(screen.queryByText('Espectador')).not.toBeInTheDocument();
  });

  test('shows spectator badge when user is SPECTATOR', () => {
    const props = {
      ...baseProps,
      roomHeaderInfo: {
        ...baseProps.roomHeaderInfo,
        userRole: 'SPECTATOR'
      }
    };
    render(<RoomTemplate {...props} />);
    expect(screen.getByText('Espectador')).toBeInTheDocument();
    expect(screen.queryByText('Jugador')).not.toBeInTheDocument();
  });

  test('shows moderator badge when user is moderator', () => {
    const props = {
      ...baseProps,
      roomHeaderInfo: {
        ...baseProps.roomHeaderInfo,
        isModerator: true
      }
    };
    render(<RoomTemplate {...props} />);
    expect(screen.getByText('Moderator')).toBeInTheDocument();
  });

  test('calls onCopyLink when button is clicked', () => {
    render(<RoomTemplate {...baseProps} />);
    fireEvent.click(screen.getByText('ðŸ“‹ Copy Room Link'));
    expect(baseProps.roomHeaderInfo.onCopyLink).toHaveBeenCalledTimes(1);
  });

  test('renders all panel components', () => {
    render(<RoomTemplate {...baseProps} />);
    expect(screen.getByTestId('participants')).toBeInTheDocument();
    expect(screen.getByTestId('story')).toBeInTheDocument();
    expect(screen.getByTestId('voting')).toBeInTheDocument();
  });

  test('only shows moderator controls when user is moderator', () => {
    // Test when not moderator
    const { rerender } = render(<RoomTemplate {...baseProps} />);
    expect(screen.queryByTestId('moderator')).not.toBeInTheDocument();
    
    // Test when is moderator
    const moderatorProps = {
      ...baseProps,
      roomHeaderInfo: {
        ...baseProps.roomHeaderInfo,
        isModerator: true
      }
    };
    rerender(<RoomTemplate {...moderatorProps} />);
    expect(screen.getByTestId('moderator')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<RoomTemplate {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});