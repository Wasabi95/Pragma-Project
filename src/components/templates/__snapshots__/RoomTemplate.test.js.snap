// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomTemplate Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="d-flex justify-content-between align-items-center mb-3"
    >
      <h2>
        Room: Test Room
      </h2>
      <div>
        <span>
          Welcome, 
          <strong>
            Test User
          </strong>
        </span>
        <span
          class="badge bg-info text-dark ms-2"
        >
          Jugador
        </span>
      </div>
    </div>
    <button
      class="btn btn-outline-primary mb-4"
      type="button"
    >
      ðŸ“‹ Copy Room Link
    </button>
    <div
      class="row"
    >
      <div
        class="col-md-4"
      >
        <div
          data-testid="participants"
        >
          Participants
        </div>
      </div>
      <div
        class="col-md-8"
      >
        <div
          data-testid="story"
        >
          Story Panel
        </div>
        <div
          data-testid="voting"
        >
          Voting Panel
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
